File: web/src/queries/platformImport.ts
// POST /api/v1/imports/platform/discover and /apply
// (internal/api/platform_import.go). The source token travels in the
// request body only and is never cached: these are mutations, so nothing
// lands in the query cache.

import { useMutation } from '@tanstack/react-query'
import { ApiError, readErrorMessage } from '../lib/apiError'

export type ImportPlatform = 'coolify' | 'dokploy' | 'caprover'

export type ImportItemStatus =
  | 'mapped'
  | 'needs-attention'
  | 'unsupported'
  | 'already-imported'
  | 'skipped'
  | 'created'
  | 'failed'

export interface PlatformImportRequest {
  platform: ImportPlatform
  url: string
  token: string
  insecure_tls?: boolean
  allow_private?: boolean
  allow_loopback?: boolean
  only?: string[]
  collision?: 'suffix' | 'skip'
}

export interface PlatformImportItem {
  kind: string
  source_id: string
  source_name: string
  target?: string
  status: ImportItemStatus
  reasons?: string[]
  manual?: string[]
}

export interface PlatformImportReport {
  platform: ImportPlatform
  items: PlatformImportItem[]
  counts: Record<string, number>
  notes?: string[]
}

async function postImport(
  action: 'discover' | 'apply',
  req: PlatformImportRequest,
): Promise<PlatformImportReport> {
  const res = await fetch(`/api/v1/imports/platform/${action}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(res, `platform import ${action} failed`),
    )
  }
  return (await res.json()) as PlatformImportReport
}

export function useDiscoverPlatformImport() {
  return useMutation({
    mutationFn: (req: PlatformImportRequest) => postImport('discover', req),
  })
}

export function useApplyPlatformImport() {
  return useMutation({
    mutationFn: (req: PlatformImportRequest) => postImport('apply', req),
  })
}

File: internal/apiclient/platform_import.go
package apiclient

import (
	"context"
	"net/http"
)

// PlatformImportRequest is the body of the platform import routes. Token
// is the source platform's credential and is sent only in this body.
type PlatformImportRequest struct {
	Platform      string   `json:"platform"`
	URL           string   `json:"url"`
	Token         string   `json:"token"`
	InsecureTLS   bool     `json:"insecure_tls,omitempty"`
	AllowPrivate  bool     `json:"allow_private,omitempty"`
	AllowLoopback bool     `json:"allow_loopback,omitempty"`
	Only          []string `json:"only,omitempty"`
	Collision     string   `json:"collision,omitempty"`
}

// PlatformImportItem is one row of a platform import report.
type PlatformImportItem struct {
	Kind       string   `json:"kind"`
	SourceID   string   `json:"source_id"`
	SourceName string   `json:"source_name"`
	Target     string   `json:"target,omitempty"`
	Status     string   `json:"status"`
	Reasons    []string `json:"reasons,omitempty"`
	Manual     []string `json:"manual,omitempty"`
}

// PlatformImportReport is the discover or apply result.
type PlatformImportReport struct {
	Platform string               `json:"platform"`
	Items    []PlatformImportItem `json:"items"`
	Counts   map[string]int       `json:"counts"`
	Notes    []string             `json:"notes,omitempty"`
}

// DiscoverPlatformImport calls POST /api/v1/imports/platform/discover.
func (c *Client) DiscoverPlatformImport(ctx context.Context, req PlatformImportRequest) (PlatformImportReport, error) {
	var out PlatformImportReport
	err := c.do(ctx, http.MethodPost, "/api/v1/imports/platform/discover", req, &out)
	return out, err
}

// ApplyPlatformImport calls POST /api/v1/imports/platform/apply.
func (c *Client) ApplyPlatformImport(ctx context.Context, req PlatformImportRequest) (PlatformImportReport, error) {
	var out PlatformImportReport
	err := c.do(ctx, http.MethodPost, "/api/v1/imports/platform/apply", req, &out)
	return out, err
}
